### PR TITLE
Refactor IConditionalAccessExpression

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/21278. Tagging @dotnet/analyzer-ioperation @dotnet/roslyn-compiler for review.

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/IConditionalAccessExpression.cs [hunk 1]
```diff
@@ -14,13 +14,13 @@ namespace Microsoft.CodeAnalysis.Semantics
...
+        /// Expression to be evaluated if <see cref="Expression"/> is non null.
         /// </summary>
-        IOperation ConditionalValue { get; }
+        IOperation WhenNotNull { get; }
         /// <summary>
-        /// Expresson that is conditionally accessed.
+        /// Expression that is conditionally accessed.
         /// </summary>
-        IOperation ConditionalInstance { get; }
+        IOperation Expression { get; }
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/IConditionalAccessExpression.cs]: >IOperation Expression { get; } [](start = 8, length = 30)

We might want to declare this property before ```WhenNotNull```.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConditionalAccessExpression.vb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        Dim o As New Object
+        o?.ToString()'BIND:"o?.ToString()"
+    End Sub
+End Class]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IConditionalAccessExpression (OperationKind.ConditionalAccessExpression, Type: System.Void) (Syntax: 'o?.ToString()')
+  Expression: ILocalReferenceExpression: o (OperationKind.LocalReferenceExpression, Type: System.Object) (Syntax: 'o')
+  WhenNotNull: IInvocationExpression (virtual Function System.Object.ToString() As System.String) (OperationKind.InvocationExpression, Type: System.String) (Syntax: '.ToString()')
+      Instance Receiver: IPlaceholderExpression (OperationKind.PlaceholderExpression, Type: System.Object) (Syntax: 'o?.ToString()')
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConditionalAccessExpression.vb]: >IPlaceholderExpression [](start = 25, length = 22)

Probably should be IConditionalAccessInstanceExpression. Do we have a separate issue tracking that?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConditionalAccessExpression.vb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        Dim c1 As C1 = Nothing
+        Dim propValue = c1?.Prop1'BIND:"c1?.Prop1"
+    End Sub
+End Class]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IConditionalAccessExpression (OperationKind.ConditionalAccessExpression, Type: System.Nullable(Of System.Int32)) (Syntax: 'c1?.Prop1')
+  Expression: ILocalReferenceExpression: c1 (OperationKind.LocalReferenceExpression, Type: C1) (Syntax: 'c1')
+  WhenNotNull: IPropertyReferenceExpression: ReadOnly Property C1.Prop1 As System.Int32 (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: '.Prop1')
+      Instance Receiver: IPlaceholderExpression (OperationKind.PlaceholderExpression, Type: C1) (Syntax: 'c1?.Prop1')
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConditionalAccessExpression.vb]: >IPlaceholderExpression [](start = 25, length = 22)

Same here.

**Review by AlekseyTs — APPROVED** [review]
LGTM. Please open an issue to track work around IConditionalAccessInstanceExpression for VB.

**Comment by 333fred** [comment]
Filed https://github.com/dotnet/roslyn/issues/21671 for the IPlaceHolderExpression issue.

File: src/Compilers/Core/Portable/Operations/IConditionalAccessExpression.cs (hunk 1 above)
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/IConditionalAccessExpression.cs]: Fixed.

**Comment by 333fred** [comment]
Test failure is https://github.com/dotnet/roslyn/issues/14592.

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_unit64_prtest  please.

**Comment by 333fred** [comment]
Ping @dotnet/roslyn-compiler for another review, @dotnet/analyzer-ioperation for a couple of reviews.

**Review by cston — APPROVED** [review]

**Comment by 333fred** [comment]
Ping @dotnet/analyzer-ioperation for a couple of reviews please.

**Comment by 333fred** [comment]
@AlekseyTs updated the doc comment to resolve https://github.com/dotnet/roslyn/issues/8861.

**Comment by AlekseyTs** [comment]
@333fred Are build failures cause by the changes in this PR?

**Comment by 333fred** [comment]
@AlekseyTs I don't think so, see my comment on https://github.com/dotnet/roslyn/pull/21643#issuecomment-325440654.

**Review by AlekseyTs — APPROVED** [review]
LGTM

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by 333fred** [comment]
Integration test failure is https://github.com/dotnet/roslyn/issues/20019.

**Pull request merged**